File: coding/solutions/src/test/java/blind75/GroupAnagramsTest.java
package blind75;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import xyz.grind.coding.blind75.GroupAnagrams;
import java.util.ArrayList;
import java.util.List;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class GroupAnagramsTest {

    List<List<String>> listOfListsInputMock(boolean testCasesRenderer) {

        List<List<String>> listOfLists = new ArrayList<>();
        List<String> firstInnerList = new ArrayList<>();
        List<String> secondInnerList = new ArrayList<>();
        List<String> thirdInnerList = new ArrayList<>();

        firstInnerList.add("eat");
        firstInnerList.add("ate");
        firstInnerList.add("tea");
        firstInnerList.add("eta");
        listOfLists.add(firstInnerList);

        if(testCasesRenderer) {
            secondInnerList.add("tba");
            secondInnerList.add("tab");
            secondInnerList.add("bat");
            listOfLists.add(secondInnerList);
        }
        thirdInnerList.add("post");
        thirdInnerList.add("pots");
        thirdInnerList.add("spot");
        thirdInnerList.add("stop");
        thirdInnerList.add("tops");
        listOfLists.add(thirdInnerList);

        return listOfLists;
    }

    @Test
    @DisplayName("Should return list of lists that has grouped anagrams that are provided from the strs input array")
    void shouldReturnArrayWithTwoIntegersThatSumUpToTarget() {
        assertAll(
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForce(new String[]{"eat","ate","tea","eta","post","pots","spot","stop","tops"}), listOfListsInputMock(false)),
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForce(new String[]{"eat","ate","tea","eta","tba","tab","bat","post","pots","spot","stop","tops"}), listOfListsInputMock(true)),
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(new String[]{"eat","ate","tea","eta","post","pots","spot","stop","tops"}), listOfListsInputMock(false)),
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(new String[]{"eat","ate","tea","eta","tba","tab","bat","post","pots","spot","stop","tops"}), listOfListsInputMock(true)),
                () -> assertEquals(GroupAnagrams.groupAnagrams(new String[]{"eat","ate","tea","eta","post","pots","spot","stop","tops"}), listOfListsInputMock(false))
        );
    }

    @Test
    @DisplayName("Should return an empty list")
    void shouldReturnEmptyArrayIfNoTwoIntegersSumUpToTarget() {
        assertAll(
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForce(new String[]{}), new ArrayList<>()),
                () -> assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(new String[]{}), new ArrayList<>()),
                () -> assertEquals(GroupAnagrams.groupAnagrams(new String[]{}), new ArrayList<>())
        );
    }

}

File: problem-solving/solutions/leetcode/blind75/src/test/kotlin/easy/GroupAnagramsTest.kt
package easy

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import xyz.grind.coding.easy.GroupAnagrams

class GroupAnagramsTest {

    private fun listOfListsInputMock(testCasesRenderer: Boolean): List<List<String>>? {
        val listOfLists: MutableList<List<String>> = ArrayList()
        val firstInnerList: MutableList<String> = ArrayList()
        val secondInnerList: MutableList<String> = ArrayList()
        val thirdInnerList: MutableList<String> = ArrayList()
        firstInnerList.add("eat")
        firstInnerList.add("ate")
        firstInnerList.add("tea")
        firstInnerList.add("eta")
        listOfLists.add(firstInnerList)
        if (testCasesRenderer) {
            secondInnerList.add("tba")
            secondInnerList.add("tab")
            secondInnerList.add("bat")
            listOfLists.add(secondInnerList)
        }
        thirdInnerList.add("post")
        thirdInnerList.add("pots")
        thirdInnerList.add("spot")
        thirdInnerList.add("stop")
        thirdInnerList.add("tops")
        listOfLists.add(thirdInnerList)
        return listOfLists
    }

    @Test
    fun `Should return list of lists that has grouped anagrams that are provided from the strs input array`() {
        Assertions.assertAll(
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForce(arrayOf("eat", "ate", "tea", "eta", "post", "pots", "spot", "stop", "tops")), listOfListsInputMock(false))
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForce(arrayOf("eat", "ate", "tea", "eta", "tba", "tab", "bat", "post", "pots", "spot", "stop", "tops")), listOfListsInputMock(true))
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(arrayOf("eat", "ate", "tea", "eta", "post", "pots", "spot", "stop", "tops")), listOfListsInputMock(false))
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(arrayOf("eat", "ate", "tea", "eta", "tba", "tab", "bat", "post", "pots", "spot", "stop", "tops")), listOfListsInputMock(true))
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagrams(arrayOf("eat", "ate", "tea", "eta", "post", "pots", "spot", "stop", "tops")), listOfListsInputMock(false))
            }
        )
    }

    @Test
    fun `Should return an empty list`() {
        Assertions.assertAll(
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForce(arrayOf()), ArrayList<String>())
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagramsBruteForceTwo(arrayOf()), ArrayList<String>())
            },
            Executable {
                assertEquals(GroupAnagrams.groupAnagrams(arrayOf()), ArrayList<String>())
            }
        )
    }
}
